pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -1242,59 +1263,72 @@ func findBridge(bridgeName string) (*netlink.Bridge, error) {
-	addr, err = lookupOrAllocateIPv4(ctx, netInstStatus,
-		status.BridgeMac)
+	var mac net.HardwareAddr
+	ipAddr := ""
+	appNum := types.AppNumInvalid
+	networkID := netInstStatus.UUID
+	log.Functionf("getUlAddrs(%d) for %s/%s\n", ifnum,
+		networkID.String(), appID.String())
+	mac, err = net.ParseMAC(ulStatus.Mac)
 	if err != nil {
-		log.Errorf("getUlAddrs: Bridge IP address allocation failed %s\n", err)
-		return bridgeIPAddr, appIPAddr, err
+		errStr := fmt.Sprintf("getUlAddrs(%s/%s): parse Mac fail: %v\n",
+			networkID.String(), appID.String(), err)
+		log.Errorf(errStr)
+		return "", errors.New(errStr)
+	}
+	// for static IP Address
+	if ulStatus.AppIPAddr != nil {
+		ipAddr = ulStatus.AppIPAddr.String()
+		if netInstStatus.DhcpRange.Contains(ulStatus.AppIPAddr) {
+			errStr := fmt.Sprintf("getUlAddrs(%s/%s): Dhcp Range contains the static IP Address %s\n",
+				networkID.String(), appID.String(), ipAddr)
+			log.Errorf(errStr)
+			return "", errors.New(errStr)
+		}
 	} else {
-		bridgeIPAddr = addr
-	}
-
-	if status.AppIPAddr != nil {
-		// Static IP assignment case.
-		// Note that appIPAddr can be in a different subnet.
-		// Assumption is that the config specifies a gateway/router
-		// in the same subnet as the static address.
-		appIPAddr = status.AppIPAddr.String()
-		recordIPAssignment(ctx, netInstStatus, status.AppIPAddr,
-			status.Mac)
-	} else if status.Mac != "" {
-		// XXX or change type of VifInfo.Mac to avoid parsing?
-		var mac net.HardwareAddr
-		mac, err = net.ParseMAC(status.Mac)
+		appNum, err = appNumOnUNetAlloc(ctx, networkID, appID,
+			netInstStatus.DhcpRange.Size, true)
 		if err != nil {
-			log.Fatal("ParseMAC failed: ", status.Mac, err)
+			errStr := fmt.Sprintf("getUlAddrs(%s/%s): App Num get fail: %v\n",
+				networkID, appID, err)
+			log.Errorf(errStr)
+			return "", errors.New(errStr)
 		}
-		log.Functionf("getUlAddrs(%d/%d for %s) app Mac %s\n",
-			ifnum, appNum, netInstStatus.UUID.String(), mac.String())
-		addr, err = lookupOrAllocateIPv4(ctx, netInstStatus, mac)
+		ipAddr, err = lookupOrAllocateIPv4(netInstStatus, appID, appNum, mac)
 		if err != nil {
-			log.Errorf("getUlAddrs: App IP address allocation failed: %s\n", err)
-			return bridgeIPAddr, appIPAddr, err
-		} else {
-			appIPAddr = addr
+			errStr := fmt.Sprintf("getUlAddrs(%s/%s): IP Addr get fail: %v\n",
+				networkID.String(), appID.String(), err)
+			log.Errorf(errStr)
+			return "", err
 		}
 	}
-	log.Functionf("getUlAddrs(%d/%d) done %s/%s\n",
-		ifnum, appNum, bridgeIPAddr, appIPAddr)
-	return bridgeIPAddr, appIPAddr, err
+
+	addr := net.ParseIP(ipAddr)
+	// assigned IP Address must be inside subnet address range
+	if !netInstStatus.Subnet.Contains(addr) {
+		errStr := fmt.Sprintf("getUlAddrs(%s-%s): IP Address out of range(%s)",
+			networkID.String(), appID.String(), ipAddr)
+		return "", errors.New(errStr)
+	}
+
+	// whether the IP Address is already assigned to some other mac
+	if duplicateIPAddr(netInstStatus, ipAddr, mac) {
```

Comment:
Why would this happen? 
If we have a check for static IPs not being in the dynamic range (which we need for other reaons) we should never get any duplicate dynamic assignments.